### DISCUSSION TITLE
Is there any way to clone or reset the TSLexer state in the external scanner?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1283] [Q&A] [answered]
This is something the internal lexer seems to be able to do, or at least it clones its TSLexer state to give to the external lexer - then merges back the cloned TSLexer state if the external scanner returns true. I'm faced with a task that is possibly impossible without being able to have a sort of external scanner for my external scanner - I think I'm rediscovering why some parsers have separate scanner/lexer/parser stages. Is it possible to clone or reset the TSLexer state in the external scanner?

**Comment by ahlinc** [comment]
I assume you want to pre scan some input and after that scan it again and make decision in your external scanner?

**Reply by ahlinc** [reply]
@ahelwer Do you have a grammar with a reproducible test that can demo the case that is unsolvable without the need that you've described?

**Reply by ahelwer** [reply]
@ahlinc it arose after getting too deep in the weeds of overriding tree-sitter's lexing behavior, as in [this](https://github.com/tree-sitter/tree-sitter/discussions/1257#discussioncomment-1018874). After spending a week or two on this approach I've decided it's not worth it and tree-sitter's default lexing behavior works well enough; it is better to concentrate on other things.

**Reply by ahlinc** [reply]
Got it, by the way the next thread may be interested for you: https://github.com/tree-sitter/tree-sitter/discussions/1092#discussioncomment-714008

**Reply by ahelwer** [reply]
@ahlinc hah, that's a neat way to do it. I was poking around in that direction (tricking tree-sitter into calling the external scanner twice in a row while saving state) but didn't fully explore it, good to know it's possible.

**Comment by ahelwer** [comment, marked as answer]
As @ahlinc pointed out, this is similar to this issue: https://github.com/tree-sitter/tree-sitter/discussions/1092#discussioncomment-714008

The solution was to trick tree-sitter into calling the external scanner twice in a row; in the first call, the end is marked then text is read ahead & stored in a buffer as scanner state to be serialized/deserialized (while emitting a placeholder token). In the second call, the buffer is scanned to identify tokens then the actual token is emitted as usual.

This is obviously a hack but it works.
